keg_storage/backends/azure.py

```diff
@@ -283,43 +284,54 @@ def create_upload_url(self, path: str, expire: typing.Union[arrow.Arrow, datetim
             client = BlobClient.from_blob_url(url)
             client.upload_blob(data)
         """
-        return self._create_sas_url(
+        warnings.warn('create_upload_url is deprecated. Use link_to instead', DeprecationWarning)
+        return self.link_to(
             path=path,
-            sas_permissions='cw',
-            expire=expire,
-            ip=ip,
+            operation=base.ShareLinkOperation.upload,
+            expire=expire
         )
 
-    def create_download_url(self, path: str, expire: typing.Union[arrow.Arrow, datetime],
-                            ip: typing.Optional[str] = None):
+    def create_download_url(self, path: str, expire: typing.Union[arrow.Arrow, datetime]):
         """
         Create an SAS URL that can be used to download a blob without any additional authentication.
         This url may be accessed directly to download the blob:
 
             requests.get(url)
         """
-        return self._create_sas_url(
+        warnings.warn('create_download_url is deprecated. Use link_to instead', DeprecationWarning)
+        return self.link_to(
             path=path,
-            sas_permissions='r',
-            expire=expire,
-            ip=ip,
+            operation=base.ShareLinkOperation.download,
+            expire=expire
         )
 
-    def _create_sas_url(self, path: str, sas_permissions: str,
-                        expire: typing.Union[arrow.Arrow, datetime],
-                        ip: typing.Optional[str] = None):
+    def link_to(
+            self,
+            path: str,
+            operation: typing.Union[base.ShareLinkOperation, str],
+            expire: typing.Union[arrow.Arrow, datetime],
+    ) -> str:
         if not self.account_url:
             raise ValueError('Cannot create a SAS URL without account credentials')
         path = self._clean_path(path)
         expire = expire.datetime if isinstance(expire, arrow.Arrow) else expire
+
+        operation = base.ShareLinkOperation.as_operation(operation)
```

Comment:
If we want to coerce from a string to an operation, we should check that it is a string and not pass the union here.